### DISCUSSION TITLE
What's with the LLVM version numbers in Emscripten

### DISCUSSION THREAD

**Discussion by MarkCallow** [emscripten-core/emscripten#17549] [Q&A] [answered]
Sometime in the last month my project started getting a _lot_ of `-Wdeprecated-builtins` warnings from Emscripten. In the most recent successful build in my CI Emscripten reported

```
clang version 15.0.0 (https://github.com/llvm/llvm-project 7effcbda49ba32991b8955821b8fdbd4f8f303e2)
```

After the warnings (the oldest build in my CI with the warnings is from 18 days ago. Emscripten reports
```
clang version 15.0.0 (https://github.com/llvm/llvm-project 17e4c217b66305e60657a48f10fe3c428c2fe4d2)
```

The reported version numbers are the same!!!

Thinking this failure to change version numbers despite a significant change must be an LLVM issue I went to their GitHub repo listed above only to discovered that 15.0.0 is not released yet! The most recent release is 15.0.0-rc2 released only 2 days ago. The previous release is 14.0.6.

I have not found a way to search for a matching commit on GitHub so I've not been able to track down the commits listed above.

What is going on here:

1. Is Emscripten using its own LLVM version numbers?
2. Is Emscripten pulling the latest commit from LLVM and using that so its users get to suffer their development pains?

Either one is bad.

I'm using the Docker image with `-name emscripten` so using emscripten:latest. I thought this was the latest _release_ of Emscripten. If so, why is it apparently using an in-development version of LLVM? How can I tell Docker to use the latest version of Emscripten that contains a released LLVM?

The fundamental issue I have here is that I'm setting source properties on some files o turn off their warnings. The files are in submodules hence I don't want to modify them.  I have to base the property settings on the reported compiler version. If I blindly set, e.g., `-Wno-deprecated-builtins` and an older compiler version is used, I get an unknown warning warning!

**Comment by tlively** [comment, marked as answer]
Emscripten continuously pulls in tip-of-tree LLVM, so it picks up LLVM changes before they reach an LLVM release. However, the LLVM version numbers are only incremented at each release. For example, when LLVM 15 is released, the tip-of-tree version number that Emscripten pulls in will be changed to 16.

> How can I tell Docker to use the latest version of Emscripten that contains a released LLVM?

Nothing changes about the Emscripten release process when there is an LLVM release, so in general there will not be an Emscripten release that uses the exact commit of the LLVM release.

> I have not found a way to search for a matching commit on GitHub so I've not been able to track down the commits listed above.

You can use the src/release-info.py script from https://chromium.googlesource.com/emscripten-releases/ to match Emscripten releases with commits in the LLVM, Binaryen, and Emscripten repos. You can also do it manually by looking at the `emscripten-releases-tags.json` file from emsdk, the looking up the given hash for your Emscripten version in the emscripten-releases repo, then looking in the DEPS file to see what LLVM, Binaryen, and Emscripten hashes that release used.

**Reply by MarkCallow** [reply]
Thank you for the answer which corresponds to my no. 2. It explains why I have been seeing LLVM 15 for so long now in Emscripten even though the first release candidate was only 3 days ago. Unfortunately this scheme exposes Emscripten users to new warnings and potentially incompatible changes that occur during an LLVM development cycle without there being any change in version number to differentiate as just happened to me. 

>   >  I have not found a way to search for a matching commit on GitHub so I've not been able to track down the commits listed above.
>
> You can use the src/release-info.py script from https://chromium.googlesource.com/emscripten-releases/ to match Emscripten releases with commits in the LLVM, Binaryen, and Emscripten repos.

The LLVM version info printed by Emscripten includes the commit. My issue is how to search for that commit in the GitHub repo. I don't wish to clone LLVM to my workarea. Anyway that's off-topic for Emscripten.

**Reply by sbc100** [reply]
I normally just click on any commit in llvm, for example `https://github.com/llvm/llvm-project/commit/5f1c7e2cc5a3c07cbc2412e851a7283c1841f520`.. and then change the sha at the end of the URL to any SHA1 you want to see.

**Reply by sbc100** [reply]
Yes, sadly emscripten users are exposed to tip of tree LLVM and there is not really any way to avoid that today.  Emscripten tends to depends on very recent changes to llvm / clang / lld for many of its features.   You can try using the last stable release of llvm with emscripten and depending on what features you use it might just work.

For more on this, and our plans to make it better see: https://github.com/emscripten-core/emscripten/issues/11362
